### PR TITLE
[Gateway] Add missing stake event

### PR DESCRIPTION
**Pull request by bryanchriswhite**
## Summary

- Adds and emits a new `EventGatewayStaked` event to the gateway module.
- Refactors `EventGatewayUnstaked` to contain the entire gateway and session end height.
- Adds the shared module keeper as a gateway keeper dependency.

## Issue

- #858

## Type of change

Select one or more from the following:

- [x] New feature, functionality or library
- [ ] Consensus breaking; add the `consensus-breaking` label if so. See #791 for details
- [ ] Bug fix
- [ ] Code health or cleanup
- [ ] Documentation
- [ ] Other (specify)

## Testing

- [ ] **Documentation**: `make docusaurus_start`; only needed if you make doc changes
- [ ] **Unit Tests**: `make go_develop_and_test`
- [ ] **LocalNet E2E Tests**: `make test_e2e`
- [ ] **DevNet E2E Tests**: Add the `devnet-test-e2e` label to the PR.

## Sanity Checklist

- [x] I have tested my changes using the available tooling
- [x] I have commented my code
- [x] I have performed a self-review of my own code; both comments & source code
- [ ] I create and reference any new tickets, if applicable
- [ ] I have left TODOs throughout the codebase, if applicable

### CODE REVIEW

File: x/gateway/keeper/msg_server_stake_gateway.go [hunk 1]
```diff
@@ -96,6 +96,14 @@ func (k msgServer) StakeGateway(
 	k.SetGateway(ctx, gateway)
 	logger.Info(fmt.Sprintf("Successfully updated stake for gateway: %+v", gateway))
 
+	sdkCtx := sdk.UnwrapSDKContext(ctx)
+	gatewayStakedEvent := &types.EventGatewayStaked{Address: gateway.GetAddress()}
+	if eventErr := sdkCtx.EventManager().EmitTypedEvent(gatewayStakedEvent); eventErr != nil {
+		err = types.ErrGatewayEmitEvent.Wrapf("(%+v): %s", gatewayStakedEvent, err)
+		logger.Error(err.Error())
+		return nil, err
```
Comment (bryanchriswhite) [inline x/gateway/keeper/msg_server_stake_gateway.go:104]:
```suggestion
		return nil, status.Error(codes.Internal, err.Error())
```

File: proto/poktroll/gateway/event.proto [hunk 2]
```diff
@@ -7,7 +7,12 @@ option (gogoproto.stable_marshaler_all) = true;
 import "cosmos_proto/cosmos.proto";
 import "gogoproto/gogo.proto";
 
-// GatewayStaked defines the event emitted when a gateway has been unstaked.
+// GatewayStaked is emitted when a gateway is staked or up-staked.
+message EventGatewayStaked {
+  string address = 1 [(cosmos_proto.scalar) = "cosmos.AddressString"]; // The Bech32 address of the gateway.
```
Comment (bryanchriswhite) [inline proto/poktroll/gateway/event.proto:12]: Replace `address` with the entire `gateway` for consistency.

File: proto/poktroll/gateway/event.proto [hunk 3]
```diff
@@ -7,7 +7,12 @@ option (gogoproto.stable_marshaler_all) = true;
...
+}
+
+// GatewayUnstaked defines the event emitted when a gateway has been unstaked.
 message EventGatewayUnstaked {
   string address = 1 [(cosmos_proto.scalar) = "cosmos.AddressString"]; // The Bech32 address of the gateway the application has changed their delegation of.
```
Comment (bryanchriswhite) [inline proto/poktroll/gateway/event.proto:17]: Replace `address` with `gateway` and add `session_end_height` field.

File: proto/poktroll/gateway/event.proto (hunk 2 above)
Comment (red-0ne) [inline proto/poktroll/gateway/event.proto:12]: Does it make sense to emit the whole gateway type? I think informing about its stake is important too.

**Review by red-0ne — COMMENTED** [review]
There's one change request but otherwise looks good.

File: x/application/keeper/auto_undelegate.go [hunk 4]
```diff
@@ -48,29 +47,25 @@ func (k Keeper) EndBlockerAutoUndelegateFromUnstakedGateways(ctx sdk.Context) er
 	return nil
 }
 
-// getUnstakeGatewayEvents returns the addresses of the gateways that were
+// getUnstakedGateways returns the addresses of the gateways that were
 // unstaked in the block.
```
Comment (bryanchriswhite) [inline x/application/keeper/auto_undelegate.go:51]:
```suggestion
// getUnstakedGateways returns the gateways which were unstaked in the given tx events.
```

File: proto/poktroll/application/event.proto [hunk 5]
```diff
@@ -19,6 +19,8 @@ enum ApplicationUnbondingReason {
 // EventApplicationStaked is emitted when an application is staked or up-staked.
 message EventApplicationStaked {
   poktroll.application.Application application = 1 [(gogoproto.jsontag) = "application"];
+  // The end height of the session in which the transfer began.
```
Comment (red-0ne) [inline proto/poktroll/application/event.proto]:
```suggestion
  // The end height of the session in which the staking occurs.
```

File: proto/poktroll/application/event.proto [hunk 6]
```diff
@@ -69,6 +79,8 @@ message EventTransferError {
 message EventApplicationUnbondingBegin {
   poktroll.application.Application application = 1 [(gogoproto.jsontag) = "application"];
   ApplicationUnbondingReason reason = 2 [(gogoproto.jsontag) = "reason"];
+  // The end height of the session in which the transfer began.
```
Comment (red-0ne) [inline proto/poktroll/application/event.proto]:
```suggestion
  // The end height of the session in which the unbonding began.
```

File: proto/poktroll/application/event.proto [hunk 7]
```diff
@@ -77,11 +89,15 @@ message EventApplicationUnbondingBegin {
 message EventApplicationUnbondingEnd {
   poktroll.application.Application application = 1 [(gogoproto.jsontag) = "application"];
   ApplicationUnbondingReason reason = 2 [(gogoproto.jsontag) = "reason"];
+  // The end height of the session in which the transfer began.
```
Comment (red-0ne) [inline proto/poktroll/application/event.proto]:
```suggestion
  // The end height of the session in which the unbonding ended.
```

File: proto/poktroll/application/event.proto [hunk 8]
```diff
@@ -77,11 +89,15 @@ message EventApplicationUnbondingBegin {
...
+  int64 session_end_height = 3 [(gogoproto.jsontag) = "session_end_height"];
 }
 
 // EventApplicationUnbondingCanceled is emitted when an application which was unbonding
 // successfully (re-)stakes before the unbonding period has elapsed. An EventApplicationStaked
 // event will also be emitted immediately after this event.
 message EventApplicationUnbondingCanceled {
   poktroll.application.Application application = 1 [(gogoproto.jsontag) = "application"];
+  // The end height of the session in which the transfer began.
```
Comment (red-0ne) [inline proto/poktroll/application/event.proto]:
```suggestion
  // The end height of the session in which the unbonding got canceled.
```

File: x/application/keeper/application.go [hunk 9]
```diff
@@ -39,6 +39,12 @@ func (k Keeper) GetApplication(
 		app.PendingUndelegations = make(map[uint64]types.UndelegatingGatewayList)
 	}
 
+	// Ensure that the DelegateeGatewayAddresses is an empty slice and not nil
+	// when unmarshalling an app that has no delegations.
+	if app.DelegateeGatewayAddresses == nil {
+		app.DelegateeGatewayAddresses = make([]string, 0)
```
Comment (red-0ne) [inline x/application/keeper/application.go]: :+1:

**Review by red-0ne — APPROVED** [review]
Some NITs on the comments, but LGTM.

Preemptively approving

File: proto/poktroll/application/event.proto (hunk 8 above)
Comment (bryanchriswhite) [reply proto/poktroll/application/event.proto]: :pray:

File: proto/poktroll/application/event.proto (hunk 7 above)
Comment (bryanchriswhite) [reply proto/poktroll/application/event.proto]: :sweat_smile:  :pray:

File: proto/poktroll/application/event.proto (hunk 5 above)
Comment (bryanchriswhite) [reply proto/poktroll/application/event.proto]: Not sure how this diff got here. I've ported this comment to #867: https://github.com/pokt-network/poktroll/pull/867/files#r1806125103

File: proto/poktroll/application/event.proto (hunk 6 above)
Comment (bryanchriswhite) [reply proto/poktroll/application/event.proto]: https://github.com/pokt-network/poktroll/pull/867/files#r1806130190

File: proto/poktroll/application/event.proto (hunk 8 above)
Comment (bryanchriswhite) [reply proto/poktroll/application/event.proto]: https://github.com/pokt-network/poktroll/pull/867/files#r1806131207

File: proto/poktroll/application/event.proto (hunk 7 above)
Comment (bryanchriswhite) [reply proto/poktroll/application/event.proto]: https://github.com/pokt-network/poktroll/pull/867/files#r1806130769

File: x/gateway/keeper/msg_server_unstake_gateway.go [hunk 10]
```diff
@@ -58,7 +58,11 @@ func (k msgServer) UnstakeGateway(
...
 
-	err = ctx.EventManager().EmitTypedEvent(&types.EventGatewayUnstaked{Address: msg.Address})
+	sessionEndHeight := k.sharedKeeper.GetSessionEndHeight(ctx, ctx.BlockHeight())
+	err = ctx.EventManager().EmitTypedEvent(&types.EventGatewayUnstaked{
+		Gateway:          &gateway,
+		SessionEndHeight: sessionEndHeight,
+	})
 	if err != nil {
 		logger.Error(fmt.Sprintf("failed to emit event for unstaking gateway: %v", err))
 		return nil, err
```
Comment (bryanchriswhite) [inline x/gateway/keeper/msg_server_unstake_gateway.go:68]:
```suggestion
		err = apptypes.ErrAppEmitEvent.Wrapf("(%+v): %s", gatewayUnstakedEvent, err)
		logger.Error(err.Error)
		return nil, status.Error(codes.Internal, err.Error())
```

**Comment by bryanchriswhite** [comment]
E2E tests passing locally but stalled in CI:
image

**Pull request merged**